// json-autotranslate/lib/services/manual.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var inquirer_1 = __importDefault(require("inquirer"));
var matchers_1 = require("../matchers");
var ManualTranslation = /** @class */ (function () {
    function ManualTranslation() {
        this.name = 'Manual Translation';
    }
    ManualTranslation.prototype.initialize = function (config, interpolationMatcher) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.interpolationMatcher = interpolationMatcher;
                return [2 /*return*/];
            });
        });
    };
    ManualTranslation.prototype.supportsLanguage = function () {
        return true;
    };
    ManualTranslation.prototype.translateStrings = function (strings, from, to) {
        return __awaiter(this, void 0, void 0, function () {
            var results, _i, strings_1, _a, key, value, replacements, result;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        results = [];
                        if (strings.length === 0) {
                            return [2 /*return*/, []];
                        }
                        console.log();
                        console.log("\u251C\u2500\u250C\u2500\u2500 Translatable strings:");
                        _i = 0, strings_1 = strings;
                        _b.label = 1;
                    case 1:
                        if (!(_i < strings_1.length)) return [3 /*break*/, 4];
                        _a = strings_1[_i], key = _a.key, value = _a.value;
                        replacements = matchers_1.replaceInterpolations(value, this.interpolationMatcher).replacements;
                        process.stdout.write('│ ├── ');
                        return [4 /*yield*/, inquirer_1["default"].prompt([
                                {
                                    name: 'result',
                                    message: "[" + from + " -> " + to + "] " + (key !== value ? "(" + key + ") " : '') + "\"" + value + "\":"
                                },
                            ])];
                    case 2:
                        result = _b.sent();
                        results.push({
                            key: key,
                            value: value,
                            translated: matchers_1.reInsertInterpolations(result.result, replacements)
                        });
                        _b.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4:
                        process.stdout.write("\u2502 \u2514\u2500\u2500 Done");
                        return [2 /*return*/, results];
                }
            });
        });
    };
    return ManualTranslation;
}());
exports.ManualTranslation = ManualTranslation;
